Custom-Pylint-checks.md

```diff
@@ -0,0 +1,598 @@
+
+  * **Displayed message**: The error message displayed to the user.
+
+  * **Message symbol**: An alias of the message ID that can be used wherever the ID can be used. For example, to disable a message with a code comment, you can specify either the ID or the symbol.
+
+  * **Message help**: Displayed when the user runs `pylint --help-msg` to learn more about your message.
+
+You can also set an `options` variable to specify options that can be set by the configuration file and read by your rule. See the [Pylint documentation](https://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html#writing-an-ast-checker) for details.
+
+Next, you need to write the checker code. Pylint lets you define methods that get called as it traverses the AST. Methods of the form `visit_nodename()` are called upon entering the subtree rooted at a node of type `nodename`, and methods of the form `leave_nodename()` are called upon leaving the subtree. Each method is called with a single argument--the node of type `nodename`.
+
+For this case, you can define a method `visit_const()` that will be called whenever Pylint reaches a `Const` node. Then you can check the node's value to see if it equals `"Hello, world!"`. The method looks like this:
+
+```python
+def visit_const(self, node):
+    if node.value == "Hello, world!":
+        self.add_message(
+            'hello-world', node=node)
+```
+
+That's it! After working through the two other kinds of linters, we'll discuss how to run and test your checker.
+
+If you want an alternative tutorial for writing an AST checker, take a look at [the one Pylint provides](https://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html#writing-an-ast-checker).
+
+#### Write the token checker
+
+##### Examine the tokens
+
+Fill a file `test.py` with the following code:
+
+```python
+s = "Hello, world!"
+print("Hello, world!")
+```
+
+Then convert the file code into tokens:
+
+```console
+$ python -m tokenize test.py
+0,0-0,0:            ENCODING       'utf-8'
+1,0-1,1:            NAME           's'
+1,2-1,3:            OP             '='
+1,4-1,19:           STRING         '"Hello, world!"'
+1,19-1,20:          NEWLINE        '\n'
+2,0-2,5:            NAME           'print'
+2,5-2,6:            OP             '('
+2,6-2,21:           STRING         '"Hello, world!"'
+2,21-2,22:          OP             ')'
+2,22-2,23:          NEWLINE        '\n'
+3,0-3,0:            ENDMARKER      ''
+```
+
+Notice that our `Hello, world!` strings are each their own tokens of type `STRING`.
+
+##### Design the token checker
+
+We could write a checker that raises an error whenever it finds a token that:
+
+* is of type "STRING"
+* has value "Hello, world!", ignoring quotation marks.
+
+After designing any rule, it's important to consider what benign code it will raise errors on (false positives) and what bad code it will miss (false negatives). For our rule:
+
+* False positives
```

Comment:
Why are we repeating all this stuff again? If it's common to all implementation types then maybe pull this out into an initial design section.

Perhaps the initial design section could also just explain the brief approaches for each type of checker and then you can do a final analysis of which one is best.

Reply:
The designs are  different for each rule, so I replaced the analyses for the token and raw checkers with a note that the failings are the same as the AST checker